ledger/apptxn_test.go

```diff
@@ -0,0 +1,426 @@
+         txn Accounts 1
+         tx_field AssetCloseTo
+         b skipamount
+noclose: int 10000
+         tx_field AssetAmount
+skipamount:
+         txn Accounts 1
+         tx_field AssetReceiver
+submit:  tx_submit
+`),
+	}
+
+	eval := l.nextBlock(t)
+	eval.txns(t, &asa, &app)
+	l.endBlock(t, eval)
+
+	// Would be better to pull these out of block, or at least check them.
+	asaIndex := basics.AssetIndex(1)
+	appIndex := basics.AppIndex(2)
+
+	fund := txntest.Txn{
+		Type:     "pay",
+		Sender:   addrs[0],
+		Receiver: appIndex.Address(),
+		Amount:   300000, // account min balance, optin min balance, plus fees
+		// stay under 1M, to avoid rewards complications
+	}
+
+	eval = l.nextBlock(t)
+	eval.txn(t, &fund)
+	l.endBlock(t, eval)
+
+	fundgold := txntest.Txn{
+		Type:          "axfer",
+		Sender:        addrs[0],
+		XferAsset:     asaIndex,
+		AssetReceiver: appIndex.Address(),
+		AssetAmount:   20000,
+	}
+
+	// Fail, because app account is not opted in.
+	eval = l.nextBlock(t)
+	eval.txn(t, &fundgold, fmt.Sprintf("asset %d missing", asaIndex))
+	l.endBlock(t, eval)
+
+	amount, in := l.asa(t, appIndex.Address(), asaIndex)
+	require.False(t, in)
+	require.Equal(t, amount, uint64(0))
+
+	optin := txntest.Txn{
+		Type:            "appl",
+		ApplicationID:   appIndex,
+		Sender:          addrs[0],
+		ApplicationArgs: [][]byte{[]byte("optin")},
+		ForeignAssets:   []basics.AssetIndex{asaIndex},
+	}
+
+	// Tell the app to opt itself in.
+	eval = l.nextBlock(t)
+	eval.txn(t, &optin)
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in)
+	require.Equal(t, amount, uint64(0))
+
+	// Now, suceed, because opted in.
+	eval = l.nextBlock(t)
+	eval.txn(t, &fundgold)
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in)
+	require.Equal(t, amount, uint64(20000))
+
+	withdraw := txntest.Txn{
+		Type:            "appl",
+		ApplicationID:   appIndex,
+		Sender:          addrs[0],
+		ApplicationArgs: [][]byte{[]byte("withdraw")},
+		ForeignAssets:   []basics.AssetIndex{asaIndex},
+		Accounts:        []basics.Address{addrs[0]},
+	}
+	eval = l.nextBlock(t)
+	eval.txn(t, &withdraw)
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in)
+	require.Equal(t, amount, uint64(10000))
+
+	eval = l.nextBlock(t)
+	eval.txn(t, withdraw.Noted("2"))
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in) // Zero left, but still opted in
+	require.Equal(t, amount, uint64(0))
+
+	eval = l.nextBlock(t)
+	eval.txn(t, withdraw.Noted("3"), "underflow on subtracting")
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in) // Zero left, but still opted in
+	require.Equal(t, amount, uint64(0))
+
+	close := txntest.Txn{
+		Type:            "appl",
+		ApplicationID:   appIndex,
+		Sender:          addrs[0],
+		ApplicationArgs: [][]byte{[]byte("close")},
```

Comment:
It might be worth testing asset close with a nonzero balance, so you can verify the closeTo addr properly gets the assets.